cumulus/client/cirrus-executor/src/tests.rs

```diff
@@ -214,14 +214,24 @@ async fn set_new_code_should_work() {
 
     let new_runtime_wasm_blob = b"new_runtime_wasm_blob".to_vec();
 
+    let best_number = alice.client.info().best_number;
+    let primary_number = best_number + 1;
+    // Although we're processing the bundle manually, the original bundle processor still works in
+    // the meanwhile, it's possible the executor alice already processed this primary block, expecting next
+    // primary block, in which case we use a dummy primary hash instead.
+    //
+    // Nice to disable the built-in bundle processor and have a full control of the executor block
+    // production manually.
```

Comment:
This bothered me before and bothers even more now. If original bundle processor works, why do we need to call it manually? Also I don't understand why random primary hash is used here.

Reply:
Relying on the original bundle processor function purely requires mocking a regular runtime upgrade properly which takes more work, a manual bundle processor makes it much easier to test the runtime upgrade processing on the executor side, which is the main purpose of this test.

The random primary hash aims to ensure this test won't be broken because of any reason out of the test scope in any case later, because there is a chance that the executor has processed primary block B and primary block B+1 is not yet produced on the end of primary node, and it doesn't really matter in this test.